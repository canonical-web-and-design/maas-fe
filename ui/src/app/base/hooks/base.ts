import { useCallback, useEffect, useRef, useState } from "react";

import { NotificationSeverity } from "@canonical/react-components";
import type { NotificationProps } from "@canonical/react-components";
import { useDispatch, useSelector } from "react-redux";

import configSelectors from "app/store/config/selectors";
import { actions as messageActions } from "app/store/message";

/**
 * Add a message in response to a state change e.g. when something is created.
 * @param addCondition - Whether the message should be added.
 * @param cleanup - A cleanup action to fire.
 * @param message - The message to be displayed.
 * @param onMessageAdded - A function to call once the message has
                                      been displayed.
 * @param messageType - The notification type.
 */
export const useAddMessage = (
  addCondition: boolean,
  cleanup: () => { type: string },
  message: string,
  onMessageAdded?: (() => void) | null,
  messageSeverity: NotificationProps["severity"] = NotificationSeverity.INFORMATION
): void => {
  const dispatch = useDispatch();

  useEffect(() => {
    if (addCondition) {
      dispatch(messageActions.add(message, messageSeverity));
      onMessageAdded && onMessageAdded();
      dispatch(cleanup());
    }
  }, [
    addCondition,
    cleanup,
    dispatch,
    message,
    messageSeverity,
    onMessageAdded,
  ]);
};

/**
 * Set the browser window title.
 * @param title - The title to set.
 */
export const useWindowTitle = (title?: string): void => {
  const maasName = useSelector(configSelectors.maasName);
  const maasNamePart = maasName ? `${maasName} ` : "";
  const titlePart = title ? `${title} | ` : "";
  useEffect(() => {
    document.title = `${titlePart}${maasNamePart}MAAS`;
  }, [maasNamePart, titlePart]);
};

/**
 * Handle checking when a value has cycled from false to true.
 * @param value - The value to check.
 * @param onCycled - The function to call when the value changes from false to true.
 */
export const useCycled = (
  value: boolean,
  onCycled?: () => void
): [boolean, () => void] => {
  const previousValue = useRef(value);
  const [hasCycled, setHasCycled] = useState(false);
  useEffect(() => {
    if (value && !previousValue.current) {
      onCycled && onCycled();
      setHasCycled(true);
    }
    if (previousValue.current !== value) {
      previousValue.current = value;
    }
  }, [value, onCycled]);
  const resetHasCycled = useCallback(() => {
    setHasCycled(false);
    previousValue.current = false;
  }, [setHasCycled]);
  return [hasCycled, resetHasCycled];
};

/**
 * Handle displaying action progress for batched actions.
 * @param processingCount - The number of items currently being processed.
 * @param onComplete - The function to call when all the items have been processed.
 * @param hasErrors - Whether there are any item errors in state.
 * @param onError - The function to call when an error occurs.
 */
export const useProcessing = (
  processingCount: number,
  onComplete: () => void,
  hasErrors = false,
  onError: () => void
): void => {
  const processingStarted = useRef(false);
  if (processingStarted.current === false && processingCount > 0) {
    processingStarted.current = true;
  }

  // If all the items have finished processing and there are no errors, run the
  // onComplete function.
  useCycled(processingCount === 0, () => {
    if (!hasErrors) {
      onComplete();
    }
  });

  // If the items are processing and errors occur, run the onError function.
  useCycled(hasErrors, () => {
    if (processingStarted) {
      onError();
    }
  });
};

/**
 * Scroll an element into view on render.
 */
export const useScrollOnRender = <T extends HTMLElement>(): ((
  targetNode: T
) => void) => {
  const htmlRef = useRef<HTMLElement>(document.querySelector("html"));
  const onRenderRef = useCallback((targetNode) => {
    if (targetNode && htmlRef?.current) {
      const { height: targetHeight, y: targetTop } =
        targetNode.getBoundingClientRect();
      const windowTop = htmlRef.current.scrollTop;
      const windowBottom = windowTop + window.innerHeight;
      const targetBottom = targetTop + targetHeight;
      // Whether the top of the target is below the bottom of the screen.
      const topOffBottom = targetTop > windowBottom;
      // Whether the top of the target is above the top of the screen.
      const topOffTop = targetTop < windowTop;
      // Whether the top of the target is on the screen.
      const topOnScreen = !topOffTop && !topOffBottom;
      // Whether the bottom of the target is below the bottom of the screen.
      const bottomOffBottom = targetBottom > windowBottom;
      // Whether the target top is on the screen but the bottom is below the bottom.
      const targetPartiallyOffBottom = topOnScreen && bottomOffBottom;
      if (topOffBottom || topOffTop || targetPartiallyOffBottom) {
        window.scrollTo({
          top: targetTop,
          left: 0,
          behavior: "smooth",
        });
      }
    }
  }, []);
  return onRenderRef;
};
