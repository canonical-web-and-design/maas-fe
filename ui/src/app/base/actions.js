export const connectWebSocket = () => {
  return {
    type: "WEBSOCKET_CONNECT"
  };
};

export const fetchAuthUser = () => {
  return {
    type: "WEBSOCKET_SEND",
    payload: {
      actionType: "FETCH_AUTH_USER",
      message: {
        method: "user.auth_user",
        type: 0
      }
    }
  };
};
