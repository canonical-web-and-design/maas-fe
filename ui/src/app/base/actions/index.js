export { default as auth } from "./auth";
export { default as controller } from "./controller";
export { default as device } from "./device";
export { default as dhcpsnippet } from "./dhcpsnippet";
export { default as general } from "./general";
export { default as machine } from "./machine";
export { default as messages } from "./messages";
export { default as packagerepository } from "./packagerepository";
export { default as scripts } from "./scripts";
export { default as subnet } from "./subnet";
export { default as user } from "./user";
export { default as websocket } from "./websocket";
