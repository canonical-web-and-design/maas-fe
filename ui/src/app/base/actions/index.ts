export { default as auth } from "./auth";
export { default as general } from "./general";
export { default as licensekeys } from "./licensekeys";
export { default as messages } from "./messages";
export { default as scripts } from "./scripts";
export { default as status } from "./status";
export { default as websocket } from "./websocket";
