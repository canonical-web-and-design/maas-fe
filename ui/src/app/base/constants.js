/**
 * Message types defined by MAAS websocket API
 */
export const MESSAGE_TYPES = {
  REQUEST: 0,
  RESPONSE: 1,
  NOTIFY: 2,
};

/**
 * Common column sizes used in <Col> components
 */
export const COL_SIZES = {
  CARD_TITLE: 2,
  DELETE_ROW_BUTTONS: 4,
  SIDEBAR: 3,
  TOTAL: 12,
};

export default MESSAGE_TYPES;
