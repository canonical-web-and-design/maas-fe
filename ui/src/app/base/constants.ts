/**
 * Common column sizes used in <Col> components
 */
export const COL_SIZES = {
  CARD_TITLE: 2,
  TABLE_CONFIRM_BUTTONS: 4,
  SIDEBAR: 3,
  TOTAL: 12,
};

export const COLOURS = {
  CAUTION: "#F99B11",
  LIGHT: "#F7F7F7",
  LINK_FADED: "#D3E4ED",
  LINK: "#0066CC",
  NEGATIVE: "#C7162B",
  POSITIVE_FADED: "#B7CCB9",
  POSITIVE_MID: "#4DAB4D",
  POSITIVE: "#0E8420",
};
