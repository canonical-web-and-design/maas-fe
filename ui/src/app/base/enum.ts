export enum HardwareType {
  Node = 0,
  CPU = 1,
  Memory = 2,
  Storage = 3,
  Network = 4,
  GPU = 5,
}
