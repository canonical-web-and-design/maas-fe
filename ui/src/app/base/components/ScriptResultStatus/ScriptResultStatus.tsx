import { scriptStatus } from "app/base/enum";
import type {
  PartialScriptResult,
  ScriptResult,
} from "app/store/scriptresult/types";

type Props = { scriptResult: ScriptResult | PartialScriptResult };

const getTestResultsIcon = (
  result: ScriptResult | PartialScriptResult
): string => {
  switch (result.status) {
    case scriptStatus.PENDING:
      return "p-icon--pending";
    case scriptStatus.RUNNING:
    case scriptStatus.APPLYING_NETCONF:
    case scriptStatus.INSTALLING:
      return "p-icon--running";
    case scriptStatus.PASSED:
      return "p-icon--success";
    case scriptStatus.FAILED:
    case scriptStatus.ABORTED:
    case scriptStatus.DEGRADED:
    case scriptStatus.FAILED_APPLYING_NETCONF:
    case scriptStatus.FAILED_INSTALLING:
      return "p-icon--error";
    case scriptStatus.TIMEDOUT:
      return "p-icon--timed-out";
    case scriptStatus.SKIPPED:
      return "p-icon--warning";
    case scriptStatus.NONE:
      return "";
    default:
      return "p-icon--help";
  }
};

const ScriptResultStatus = ({ scriptResult }: Props): JSX.Element => (
  <div className="u-flex">
    <div>
      <i className={getTestResultsIcon(scriptResult)} />
    </div>
    <span className="u-nudge-right--small" data-test="status-name">
      {scriptResult.status_name}
    </span>
  </div>
);

export default ScriptResultStatus;
