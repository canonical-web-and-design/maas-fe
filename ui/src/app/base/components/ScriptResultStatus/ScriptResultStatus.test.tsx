import { shallow } from "enzyme";

import ScriptResultStatus from "./ScriptResultStatus";

import { scriptStatus } from "app/base/enum";
import { scriptResult as scriptResultFactory } from "testing/factories";

describe("ScriptResultStatus", () => {
  it("renders", () => {
    const scriptResult = scriptResultFactory({
      status: scriptStatus.PASSED,
      status_name: "Passed",
    });
    const wrapper = shallow(<ScriptResultStatus scriptResult={scriptResult} />);
    expect(wrapper).toMatchSnapshot();
  });
});
