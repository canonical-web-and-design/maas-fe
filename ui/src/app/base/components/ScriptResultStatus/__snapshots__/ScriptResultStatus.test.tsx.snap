// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScriptResultStatus renders 1`] = `
<div
  className="u-flex"
>
  <div>
    <i
      className="p-icon--success"
    />
  </div>
  <span
    className="u-nudge-right--small"
    data-test="status-name"
  >
    Passed
  </span>
</div>
`;
